{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "adb6e9cd-2601-4f8f-b8d9-fe9cc1377b87",
   "metadata": {},
   "source": [
    "## SCFEA \n",
    "Creating matrix to run scFEA in http://scflux.org/\n",
    "The input of scFEA is a scRNA-seq or general transcriptomics data, in which each row is one gene and each column is one sample. TPM (or CPM/FPKM) normalized data is recommended. scFEA webserver accepts comma-(.csv), space-(.txt), tab-(.txt) delimited input fills. Please make sure the input data is in a matrix form and contains row/column names. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "45dbfdee-9cef-4759-a66a-5d9e6d4ee79b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import decoupler as dc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "38f3d7ad-6f6e-4df9-a555-fc81c14a06d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"adata_solo_annotated_all.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5ca1a507-6bf2-4796-baea-559971077a8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 60459 × 11945\n",
       "    obs: '_scvi_batch', '_scvi_labels', 'batch', 'batch_id', 'cell_type', 'condition', 'is_doublet', 'is_outlier', 'is_outlier_counts', 'is_outlier_genes', 'is_outlier_mito', 'is_outlier_top_20', 'leiden', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'log1p_total_counts_mt', 'log1p_total_counts_ribo', 'n_counts', 'n_genes', 'n_genes_by_counts', 'origin', 'pct_counts_in_top_20_genes', 'pct_counts_mt', 'pct_counts_ribo', 'sample_id', 'total_counts', 'total_counts_mt', 'total_counts_ribo', 'value', 'outlier', 'mt_outlier', 'ribo_outlier'\n",
       "    var: 'ensembl_id', 'feature_types', 'gene_name', 'mt', 'ribo', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'n_counts', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'highly_variable_nbatches'\n",
       "    uns: 'cell_type_colors', 'hvg', 'is_doublet_colors', 'leiden', 'leiden_colors', 'log1p', 'neighbors', 'neighbors_uncorrected', 'origin_colors', 'sample_id_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_scAR', 'X_scVI', 'X_umap', 'X_umap_uncorrected'\n",
       "    layers: 'counts', 'denoised', 'logcounts'\n",
       "    obsp: 'connectivities', 'distances', 'neighbors_uncorrected_connectivities', 'neighbors_uncorrected_distances'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0aa7830c-41f2-4491-a2da-9216cc4e74bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_cpm = adata.copy()  # apply this to a copy so we can compare methods\n",
    "adata_cpm.raw = adata_cpm  # store a copy of the raw values before normalizing\n",
    "sc.pp.normalize_per_cell(adata_cpm, \n",
    "                         counts_per_cell_after=1e6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "246cf5f9-89d0-44da-b86b-67d5c6efa40c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdata = dc.get_pseudobulk(adata,\n",
    "                          sample_col='sample_id',\n",
    "                          groups_col=\"cell_type\",\n",
    "                          layer='counts',\n",
    "                          mode='sum',\n",
    "                          min_cells=100,\n",
    "                          min_counts=1000\n",
    "                         )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "540b9393-59e8-477c-98b4-b362f2d8acbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "scales_counts = sc.pp.normalize_total(pdata, target_sum=None, inplace=False)\n",
    "# log1p transform\n",
    "pdata.layers[\"log1p_norm\"] = sc.pp.log1p(scales_counts[\"X\"], copy=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f202eae3-6899-464c-bb72-8119f009bf72",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Step 1: Extract the log1p_norm layer\n",
    "log1p_norm_matrix = pdata.layers[\"log1p_norm\"]\n",
    "\n",
    "# Step 2: Convert to a pandas DataFrame\n",
    "# Ensure the rows (index) correspond to samples (adata.obs_names) and columns to genes (adata.var_names)\n",
    "log1p_norm_df = pd.DataFrame(log1p_norm_matrix,   \n",
    "                             index=pdata.obs_names,   # Sample names as row indices\n",
    "                             columns=pdata.var.gene_name) # Gene names as column headers\n",
    "\n",
    "# Step 3: Export to CSV\n",
    "#log1p_norm_df.T.to_csv(\"log1p_norm_matrix.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af9aed87-c7a8-4e81-8b05-2a16823fd8d3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b0a35509-555b-42f1-a0b3-3477c679dae7",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('scFEA.mouse.genes.txt', 'r') as f:\n",
    "    gene_list = f.read().splitlines()  # This creates a list of genes\n",
    "\n",
    "# Step 2: Filter the DataFrame to keep only the genes in the list\n",
    "filtered_log1p_norm_df = log1p_norm_df.T.loc[log1p_norm_df.T.index.isin(gene_list)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6659eb08-203d-43e1-832e-4e2b71c41d4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10mix1_COLON_Exhausted</th>\n",
       "      <th>10mix2_COLON_Exhausted</th>\n",
       "      <th>11mix1_COLON_Exhausted</th>\n",
       "      <th>11mix2_COLON_Exhausted</th>\n",
       "      <th>10mix1_COLON_Infg</th>\n",
       "      <th>10mix2_COLON_Infg</th>\n",
       "      <th>11mix1_COLON_Infg</th>\n",
       "      <th>11mix2_COLON_Infg</th>\n",
       "      <th>GF1_COLON_Infg</th>\n",
       "      <th>GF2_COLON_Infg</th>\n",
       "      <th>...</th>\n",
       "      <th>GF_ICI2_SLEC_Progenitor</th>\n",
       "      <th>GF_ICI2_plus_SLEC_Progenitor</th>\n",
       "      <th>10mix_ICI1_SLEC_Terminal</th>\n",
       "      <th>10mix_ICI2_SLEC_Terminal</th>\n",
       "      <th>11mix_ICI1_SLEC_Terminal</th>\n",
       "      <th>11mix_ICI2_SLEC_Terminal</th>\n",
       "      <th>GF_ICI1_SLEC_Terminal</th>\n",
       "      <th>GF_ICI1_plus_SLEC_Terminal</th>\n",
       "      <th>GF_ICI2_SLEC_Terminal</th>\n",
       "      <th>GF_ICI2_plus_SLEC_Terminal</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Dlat</th>\n",
       "      <td>3.998096</td>\n",
       "      <td>3.994420</td>\n",
       "      <td>4.015038</td>\n",
       "      <td>3.762799</td>\n",
       "      <td>4.264687</td>\n",
       "      <td>4.013443</td>\n",
       "      <td>3.873187</td>\n",
       "      <td>4.084481</td>\n",
       "      <td>4.330668</td>\n",
       "      <td>4.036375</td>\n",
       "      <td>...</td>\n",
       "      <td>4.452443</td>\n",
       "      <td>4.882694</td>\n",
       "      <td>4.759019</td>\n",
       "      <td>4.686661</td>\n",
       "      <td>4.674407</td>\n",
       "      <td>4.568219</td>\n",
       "      <td>4.789472</td>\n",
       "      <td>4.703448</td>\n",
       "      <td>4.684748</td>\n",
       "      <td>4.647597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Comt</th>\n",
       "      <td>4.364779</td>\n",
       "      <td>4.399291</td>\n",
       "      <td>4.139816</td>\n",
       "      <td>4.280936</td>\n",
       "      <td>3.938837</td>\n",
       "      <td>4.277446</td>\n",
       "      <td>4.165078</td>\n",
       "      <td>3.817793</td>\n",
       "      <td>4.561538</td>\n",
       "      <td>3.839621</td>\n",
       "      <td>...</td>\n",
       "      <td>4.972683</td>\n",
       "      <td>5.105197</td>\n",
       "      <td>5.085173</td>\n",
       "      <td>5.107409</td>\n",
       "      <td>5.181208</td>\n",
       "      <td>5.072165</td>\n",
       "      <td>5.067990</td>\n",
       "      <td>4.897975</td>\n",
       "      <td>4.943543</td>\n",
       "      <td>4.959501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dbt</th>\n",
       "      <td>3.692427</td>\n",
       "      <td>3.589845</td>\n",
       "      <td>3.429869</td>\n",
       "      <td>3.547163</td>\n",
       "      <td>3.827842</td>\n",
       "      <td>3.338206</td>\n",
       "      <td>3.123965</td>\n",
       "      <td>3.754719</td>\n",
       "      <td>3.483602</td>\n",
       "      <td>3.724522</td>\n",
       "      <td>...</td>\n",
       "      <td>3.859846</td>\n",
       "      <td>3.488495</td>\n",
       "      <td>3.399422</td>\n",
       "      <td>3.655519</td>\n",
       "      <td>3.485469</td>\n",
       "      <td>3.599457</td>\n",
       "      <td>3.470862</td>\n",
       "      <td>3.577276</td>\n",
       "      <td>3.624781</td>\n",
       "      <td>3.488300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hk2</th>\n",
       "      <td>3.346372</td>\n",
       "      <td>3.661979</td>\n",
       "      <td>3.665019</td>\n",
       "      <td>3.653081</td>\n",
       "      <td>4.383716</td>\n",
       "      <td>4.248013</td>\n",
       "      <td>4.246089</td>\n",
       "      <td>4.215906</td>\n",
       "      <td>4.475476</td>\n",
       "      <td>4.121913</td>\n",
       "      <td>...</td>\n",
       "      <td>3.980014</td>\n",
       "      <td>3.865758</td>\n",
       "      <td>3.903101</td>\n",
       "      <td>4.017314</td>\n",
       "      <td>3.290555</td>\n",
       "      <td>3.701058</td>\n",
       "      <td>3.505365</td>\n",
       "      <td>3.702725</td>\n",
       "      <td>3.977479</td>\n",
       "      <td>3.820938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cyp51</th>\n",
       "      <td>4.931952</td>\n",
       "      <td>4.868849</td>\n",
       "      <td>5.255898</td>\n",
       "      <td>5.097463</td>\n",
       "      <td>5.036140</td>\n",
       "      <td>4.753106</td>\n",
       "      <td>5.153751</td>\n",
       "      <td>5.089606</td>\n",
       "      <td>5.017214</td>\n",
       "      <td>5.061007</td>\n",
       "      <td>...</td>\n",
       "      <td>5.027122</td>\n",
       "      <td>5.245385</td>\n",
       "      <td>4.925780</td>\n",
       "      <td>5.081336</td>\n",
       "      <td>5.097423</td>\n",
       "      <td>5.070996</td>\n",
       "      <td>5.073660</td>\n",
       "      <td>5.150333</td>\n",
       "      <td>5.009543</td>\n",
       "      <td>5.080003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cpt1b</th>\n",
       "      <td>1.084212</td>\n",
       "      <td>1.859381</td>\n",
       "      <td>2.302855</td>\n",
       "      <td>1.584875</td>\n",
       "      <td>2.103416</td>\n",
       "      <td>1.983710</td>\n",
       "      <td>2.109662</td>\n",
       "      <td>2.495346</td>\n",
       "      <td>1.598803</td>\n",
       "      <td>1.801120</td>\n",
       "      <td>...</td>\n",
       "      <td>1.105228</td>\n",
       "      <td>1.823463</td>\n",
       "      <td>1.249915</td>\n",
       "      <td>1.405982</td>\n",
       "      <td>1.140862</td>\n",
       "      <td>1.591632</td>\n",
       "      <td>0.856530</td>\n",
       "      <td>1.341702</td>\n",
       "      <td>1.016268</td>\n",
       "      <td>1.133333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ak6</th>\n",
       "      <td>4.560105</td>\n",
       "      <td>4.727462</td>\n",
       "      <td>4.657283</td>\n",
       "      <td>4.356413</td>\n",
       "      <td>4.656953</td>\n",
       "      <td>4.658658</td>\n",
       "      <td>4.646770</td>\n",
       "      <td>4.313604</td>\n",
       "      <td>4.429498</td>\n",
       "      <td>4.923795</td>\n",
       "      <td>...</td>\n",
       "      <td>4.379231</td>\n",
       "      <td>4.091959</td>\n",
       "      <td>4.308120</td>\n",
       "      <td>4.668933</td>\n",
       "      <td>4.549068</td>\n",
       "      <td>4.658709</td>\n",
       "      <td>4.197059</td>\n",
       "      <td>4.321927</td>\n",
       "      <td>4.262834</td>\n",
       "      <td>4.277251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uckl1</th>\n",
       "      <td>3.828935</td>\n",
       "      <td>4.043125</td>\n",
       "      <td>4.036946</td>\n",
       "      <td>4.105483</td>\n",
       "      <td>3.973256</td>\n",
       "      <td>4.013443</td>\n",
       "      <td>4.296666</td>\n",
       "      <td>3.615213</td>\n",
       "      <td>4.220989</td>\n",
       "      <td>3.055213</td>\n",
       "      <td>...</td>\n",
       "      <td>4.272449</td>\n",
       "      <td>4.139034</td>\n",
       "      <td>4.308120</td>\n",
       "      <td>4.152883</td>\n",
       "      <td>4.293873</td>\n",
       "      <td>4.137671</td>\n",
       "      <td>4.300234</td>\n",
       "      <td>4.162763</td>\n",
       "      <td>4.204417</td>\n",
       "      <td>4.191854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hmgcs1</th>\n",
       "      <td>4.231878</td>\n",
       "      <td>4.230433</td>\n",
       "      <td>4.542140</td>\n",
       "      <td>4.485462</td>\n",
       "      <td>4.448877</td>\n",
       "      <td>3.934905</td>\n",
       "      <td>4.646770</td>\n",
       "      <td>4.769176</td>\n",
       "      <td>4.640777</td>\n",
       "      <td>4.121913</td>\n",
       "      <td>...</td>\n",
       "      <td>4.624931</td>\n",
       "      <td>4.560517</td>\n",
       "      <td>4.330899</td>\n",
       "      <td>4.413214</td>\n",
       "      <td>4.371746</td>\n",
       "      <td>4.360596</td>\n",
       "      <td>4.484377</td>\n",
       "      <td>4.446547</td>\n",
       "      <td>4.389260</td>\n",
       "      <td>4.370234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pnp</th>\n",
       "      <td>5.195186</td>\n",
       "      <td>5.366804</td>\n",
       "      <td>5.482263</td>\n",
       "      <td>5.416405</td>\n",
       "      <td>5.749215</td>\n",
       "      <td>5.764134</td>\n",
       "      <td>6.008187</td>\n",
       "      <td>5.970747</td>\n",
       "      <td>5.402849</td>\n",
       "      <td>5.363028</td>\n",
       "      <td>...</td>\n",
       "      <td>5.408026</td>\n",
       "      <td>5.423269</td>\n",
       "      <td>5.400534</td>\n",
       "      <td>5.534292</td>\n",
       "      <td>5.584512</td>\n",
       "      <td>5.626914</td>\n",
       "      <td>5.343360</td>\n",
       "      <td>5.388087</td>\n",
       "      <td>5.415585</td>\n",
       "      <td>5.466923</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>411 rows × 87 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           10mix1_COLON_Exhausted  10mix2_COLON_Exhausted  \\\n",
       "gene_name                                                   \n",
       "Dlat                     3.998096                3.994420   \n",
       "Comt                     4.364779                4.399291   \n",
       "Dbt                      3.692427                3.589845   \n",
       "Hk2                      3.346372                3.661979   \n",
       "Cyp51                    4.931952                4.868849   \n",
       "...                           ...                     ...   \n",
       "Cpt1b                    1.084212                1.859381   \n",
       "Ak6                      4.560105                4.727462   \n",
       "Uckl1                    3.828935                4.043125   \n",
       "Hmgcs1                   4.231878                4.230433   \n",
       "Pnp                      5.195186                5.366804   \n",
       "\n",
       "           11mix1_COLON_Exhausted  11mix2_COLON_Exhausted  10mix1_COLON_Infg  \\\n",
       "gene_name                                                                      \n",
       "Dlat                     4.015038                3.762799           4.264687   \n",
       "Comt                     4.139816                4.280936           3.938837   \n",
       "Dbt                      3.429869                3.547163           3.827842   \n",
       "Hk2                      3.665019                3.653081           4.383716   \n",
       "Cyp51                    5.255898                5.097463           5.036140   \n",
       "...                           ...                     ...                ...   \n",
       "Cpt1b                    2.302855                1.584875           2.103416   \n",
       "Ak6                      4.657283                4.356413           4.656953   \n",
       "Uckl1                    4.036946                4.105483           3.973256   \n",
       "Hmgcs1                   4.542140                4.485462           4.448877   \n",
       "Pnp                      5.482263                5.416405           5.749215   \n",
       "\n",
       "           10mix2_COLON_Infg  11mix1_COLON_Infg  11mix2_COLON_Infg  \\\n",
       "gene_name                                                            \n",
       "Dlat                4.013443           3.873187           4.084481   \n",
       "Comt                4.277446           4.165078           3.817793   \n",
       "Dbt                 3.338206           3.123965           3.754719   \n",
       "Hk2                 4.248013           4.246089           4.215906   \n",
       "Cyp51               4.753106           5.153751           5.089606   \n",
       "...                      ...                ...                ...   \n",
       "Cpt1b               1.983710           2.109662           2.495346   \n",
       "Ak6                 4.658658           4.646770           4.313604   \n",
       "Uckl1               4.013443           4.296666           3.615213   \n",
       "Hmgcs1              3.934905           4.646770           4.769176   \n",
       "Pnp                 5.764134           6.008187           5.970747   \n",
       "\n",
       "           GF1_COLON_Infg  GF2_COLON_Infg  ...  GF_ICI2_SLEC_Progenitor  \\\n",
       "gene_name                                  ...                            \n",
       "Dlat             4.330668        4.036375  ...                 4.452443   \n",
       "Comt             4.561538        3.839621  ...                 4.972683   \n",
       "Dbt              3.483602        3.724522  ...                 3.859846   \n",
       "Hk2              4.475476        4.121913  ...                 3.980014   \n",
       "Cyp51            5.017214        5.061007  ...                 5.027122   \n",
       "...                   ...             ...  ...                      ...   \n",
       "Cpt1b            1.598803        1.801120  ...                 1.105228   \n",
       "Ak6              4.429498        4.923795  ...                 4.379231   \n",
       "Uckl1            4.220989        3.055213  ...                 4.272449   \n",
       "Hmgcs1           4.640777        4.121913  ...                 4.624931   \n",
       "Pnp              5.402849        5.363028  ...                 5.408026   \n",
       "\n",
       "           GF_ICI2_plus_SLEC_Progenitor  10mix_ICI1_SLEC_Terminal  \\\n",
       "gene_name                                                           \n",
       "Dlat                           4.882694                  4.759019   \n",
       "Comt                           5.105197                  5.085173   \n",
       "Dbt                            3.488495                  3.399422   \n",
       "Hk2                            3.865758                  3.903101   \n",
       "Cyp51                          5.245385                  4.925780   \n",
       "...                                 ...                       ...   \n",
       "Cpt1b                          1.823463                  1.249915   \n",
       "Ak6                            4.091959                  4.308120   \n",
       "Uckl1                          4.139034                  4.308120   \n",
       "Hmgcs1                         4.560517                  4.330899   \n",
       "Pnp                            5.423269                  5.400534   \n",
       "\n",
       "           10mix_ICI2_SLEC_Terminal  11mix_ICI1_SLEC_Terminal  \\\n",
       "gene_name                                                       \n",
       "Dlat                       4.686661                  4.674407   \n",
       "Comt                       5.107409                  5.181208   \n",
       "Dbt                        3.655519                  3.485469   \n",
       "Hk2                        4.017314                  3.290555   \n",
       "Cyp51                      5.081336                  5.097423   \n",
       "...                             ...                       ...   \n",
       "Cpt1b                      1.405982                  1.140862   \n",
       "Ak6                        4.668933                  4.549068   \n",
       "Uckl1                      4.152883                  4.293873   \n",
       "Hmgcs1                     4.413214                  4.371746   \n",
       "Pnp                        5.534292                  5.584512   \n",
       "\n",
       "           11mix_ICI2_SLEC_Terminal  GF_ICI1_SLEC_Terminal  \\\n",
       "gene_name                                                    \n",
       "Dlat                       4.568219               4.789472   \n",
       "Comt                       5.072165               5.067990   \n",
       "Dbt                        3.599457               3.470862   \n",
       "Hk2                        3.701058               3.505365   \n",
       "Cyp51                      5.070996               5.073660   \n",
       "...                             ...                    ...   \n",
       "Cpt1b                      1.591632               0.856530   \n",
       "Ak6                        4.658709               4.197059   \n",
       "Uckl1                      4.137671               4.300234   \n",
       "Hmgcs1                     4.360596               4.484377   \n",
       "Pnp                        5.626914               5.343360   \n",
       "\n",
       "           GF_ICI1_plus_SLEC_Terminal  GF_ICI2_SLEC_Terminal  \\\n",
       "gene_name                                                      \n",
       "Dlat                         4.703448               4.684748   \n",
       "Comt                         4.897975               4.943543   \n",
       "Dbt                          3.577276               3.624781   \n",
       "Hk2                          3.702725               3.977479   \n",
       "Cyp51                        5.150333               5.009543   \n",
       "...                               ...                    ...   \n",
       "Cpt1b                        1.341702               1.016268   \n",
       "Ak6                          4.321927               4.262834   \n",
       "Uckl1                        4.162763               4.204417   \n",
       "Hmgcs1                       4.446547               4.389260   \n",
       "Pnp                          5.388087               5.415585   \n",
       "\n",
       "           GF_ICI2_plus_SLEC_Terminal  \n",
       "gene_name                              \n",
       "Dlat                         4.647597  \n",
       "Comt                         4.959501  \n",
       "Dbt                          3.488300  \n",
       "Hk2                          3.820938  \n",
       "Cyp51                        5.080003  \n",
       "...                               ...  \n",
       "Cpt1b                        1.133333  \n",
       "Ak6                          4.277251  \n",
       "Uckl1                        4.191854  \n",
       "Hmgcs1                       4.370234  \n",
       "Pnp                          5.466923  \n",
       "\n",
       "[411 rows x 87 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_log1p_norm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "aff1a075-bc78-4a16-a90a-5b352c260fe1",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_log1p_norm_df.to_csv(\"log1p_norm_matrix.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eeaad600-6576-4e1a-a239-d9a8d68065f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_log1p_norm_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
